[countbuild][xplat/caffe2] contbuild with sanitizers

pavithranrao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61724

To improve the stability of xplat/caffe2 code, we are enabling sanitizers (asan, tsan, ubsan) on contbuild.

Differential Revision: [D29348099](https://our.internmc.facebook.com/intern/diff/D29348099/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29348099/)!

albanD commented on torch/csrc/autograd/functions/basic_ops.h:
```diff
@@ -42,7 +42,7 @@ struct TORCH_API DelayedError : public Node {
     : msg(std::move(msg)) {
     // NOLINTNEXTLINE(clang-analyzer-deadcode.DeadStores,clang-diagnostic-unused-variable)
-    for (const auto i : c10::irange(num_inputs))
+    for ([[maybe_unused]] const auto i : c10::irange(num_inputs))
```
Should we just rewrite this loop to avoid the unused variable issue?

albanD commented on c10/macros/Macros.h:
```diff
@@ -60,6 +60,23 @@
+// for clang
+#if defined(__has_feature)
+#if ((__has_feature(thread_sanitizer)))
+#define C10_NO_TSAN __attribute__((no_sanitize("thread")))
```
These macros are used in the codebase?

pavithranrao replied:
will remove macro

albanD reviewed (approved):
LGTM
Thanks for the update.
Should the comment annotation be updated btw?

pavithranrao committed:
[countbuild][xplat/caffe2] contbuild with sanitizers (#61724)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/61724

To improve the stability of xplat/caffe2 code, we are enabling sanitizers (asan, tsan, ubsan) on contbuild.
ghstack-source-id: 134339882

Test Plan:
```
buck test --show-output --flagfile fbsource//fbcode/mode/dev-asan --config fbsource.sanitizer=address fbsource//xplat/pytorch_models/build/pytorch_model_test/v13:body_tracking_v124_test
clang-9: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]

Downloaded 0/7 artifacts, 0.00 bytes, 100.0% cache miss
Building: finished in 14.5 sec (100%) 4538/4538 jobs, 4 updated
  Total time: 14.5 sec
Testing: finished in 1.1 sec (1 PASS/0 FAIL)
RESULTS FOR //xplat/pytorch_models/build/pytorch_model_test/v13:body_tracking_v124_test
PASS      1.0s  1 Passed   0 Skipped   0 Failed   //xplat/pytorch_models/build/pytorch_model_test/v13:body_tracking_v124_test
TESTS PASSED
```

```
buck test --show-output --flagfile fbsource//fbcode/mode/dev-tsan --config fbsource.sanitizer=thread fbsource//xplat/pytorch_models/build/ads_mai_test_train/v4:model_test
clang-9: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]

Downloaded 3/19 artifacts, 88.30 Kbytes, 66.7% cache miss
Building: finished in 24.0 sec (100%) 4609/4609 jobs, 9 updated
  Total time: 24.9 sec
Testing: finished in 0.9 sec (1 PASS/0 FAIL)
RESULTS FOR //xplat/pytorch_models/build/ads_mai_test_train/v4:model_test
PASS     808ms  1 Passed   0 Skipped   0 Failed   //xplat/pytorch_models/build/ads_mai_test_train/v4:model_test
TESTS PASSED
````

Reviewed By: dhruvbird, albanD

Differential Revision: D29348099

fbshipit-source-id: 1b6ccbc4a9d3b04fd218b1e5ac5a696cf50602e0
